lib/urlwatch/reporters.py

```diff
@@ -76,6 +76,35 @@
 WDIFF_REMOVED_RE = r'[\[][-].*?[-][]]'
 
 
+@functools.lru_cache(maxsize=1)
+def cached_unified_diff(old_data, new_data, timestamp_old, timestamp_new, comparison_filter):
```

Comment:
This could use a unit test or three :)

Reply:
Hmm... but there are no errors if we don't test! ;)

OK great idea.

Reply:
OK, I took a look at the (undocumented) files in the `test` folder and have absolutely no idea on how to do unit testing of this job filter. I believe it would require running a job with a specified filter to set the baseline, then running another job on the same source with a different filter that has been modified to generate a diff (e.g. through the use of re.sub or grep/grepi), capture its output, and see if it matches expectations.  

I couldn't find anything in there remotely resembling this and I cannot understand enough about the structure of the program the nesting inside a nest inside a nest to write one from scratch -- nor do I have the time to do so.

I am afraid it's something someone else will have to tackle.

Reply:
The `cached_unified_diff()` doesn't have anything to do with jobs.

Create a function `test_cache_unified_diff()` in a new file `test_reporters.py` and add something like this:

```python
from urlwatch.reporters import cache_unified_diff
from nose.tools import eq_

def test_cache_unified_diff_normal():
    input_text_a = 'a'
    input_text_b = 'b'
    expected_diff_output = "..."  # put here the result you expect the function to return
    eq_(cache_unified_diff(input_text_a, input_text_b, '', '', None), expected_diff_output)
```

And then add tests for other cases, e.g.:

- changed line with "additions" filter
- changed line with "deletions" filter
- "additions" filter with more than 75% deletions
- "additions" filter and lines were only deleted
- "deletions" filter and lines were only added

Reply:
Thanks!!! Unit testing written